community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/slurm_files/scripts/slurmsync.py

```diff
@@ -54,21 +54,91 @@
 TOT_REQ_CNT = 1000
 _MAINTENANCE_SBATCH_SCRIPT_PATH = dirs.custom_scripts / "perform_maintenance.sh"
 
-NodeStatus = Enum(
-    "NodeStatus",
-    (
-        "orphan",
-        "power_down",
-        "preempted",
-        "restore",
-        "resume",
-        "terminated",
-        "unbacked",
-        "unchanged",
-        "unknown",
-    ),
-)
+class NodeAction(Protocol):
+    def apply(self, nodes: List[str], hostlist: List[str], count: int) -> None:
+        ...
+
+    def __hash__(self):
+        return hash(type(self))
```

Comment:
We can't just use hash of type, consider following action:

```py
class SomeAction(NodeAction):
  def __init__(sefl, reason):
    self.reason = reason
    
  def apply(nodes: List[str]):
    run(f"scontrol .... reason={self.reason}")
```

This should be true:
```py
SomeAction("why not").__hash__() != SomeAction("just").__hash__()
```

Reply:
This would only apply to `NodeActionDown` at the moment right? Other actions use an instance of `NodeActionDown` will always have the same `reason` and hence should give the same hash. Updated `NodeActionDown`'s hash function to account for reason attribute.